File: stack-typescript/src/tools/reject-timesheet.ts
import { tool } from "ai";
import { z } from "zod";
import { authorize, tenantFor } from "../policy/approval.js";
import type { ToolError } from "../types.js";
import { upstreamFor } from "../upstream/index.js";
import { callerFromOptions, callerSchema } from "./context.js";

export async function executeReject(caller: Parameters<typeof authorize>[0], timesheet_id: string, reason: string) {
  const baseline = authorize(caller, "reject", null);
  if (!baseline.allowed) return baseline;
  const tenant = tenantFor(caller);
  const upstream = upstreamFor(caller);
  const row = await upstream.detail(tenant, timesheet_id);
  if (!row) return { error: "not_found", message: "Timesheet was not found." } satisfies ToolError;
  const verdict = authorize(caller, "reject", row);
  if (!verdict.allowed && !(verdict.error === "requires_human_approval" && caller.humanDecisionFor === timesheet_id)) return verdict;
  return upstream.reject(tenant, timesheet_id, caller.sub, reason);
}

export const rejectTimesheet = tool({
  description: "Reject one timesheet on behalf of the current user with a reason the worker will see. Same authorization rules as approve_timesheet. Always requires a reason.",
  inputSchema: z.object({ timesheet_id: z.string(), reason: z.string().min(3).max(500) }).strict(),
  contextSchema: callerSchema,
  execute: ({ timesheet_id, reason }, options) => executeReject(callerFromOptions(options), timesheet_id, reason),
});

File: stack-dotnet/src/Tools/RejectTimesheet.cs
using System.ComponentModel;
using StackDotnet.Policy;

namespace StackDotnet.Tools;

public static class RejectTimesheet
{
    [Description("Reject one timesheet on behalf of the current user with a reason the worker will see. Same authorization rules as approve_timesheet. Always requires a reason.")]
    public static async Task<object> Execute(ToolServices services, CallerContext caller, HumanDecisionContext humanDecision, [Description("Timesheet identifier.")] string timesheet_id, [Description("Reason the worker will see, from 3 through 500 characters.")] string reason, CancellationToken cancellationToken = default)
    {
        var invocation = Authorizer.Authorize(caller, AuthorizationAction.Reject, null);
        if (!invocation.Allowed) return new ToolError(invocation.Error!, invocation.Message!);
        if (reason.Length is < 3 or > 500) return new ToolError("validation", "reason must be between 3 and 500 characters");
        var org = services.OrgName(caller);
        var policyCaller = services.PolicyCaller(caller);
        var row = await services.Cws.GetAsync(org, timesheet_id, cancellationToken);
        if (row is null) return new ToolError("not_found", "Timesheet was not found.");
        var authorization = Authorizer.Authorize(policyCaller, AuthorizationAction.Reject, row, ApprovalPolicy.For(org));
        if (!authorization.Allowed && !(authorization.Error == "requires_human_approval" && humanDecision.ApprovedIds.Contains(timesheet_id)))
            return new ToolError(authorization.Error!, authorization.Message!);
        return await services.Cws.RejectAsync(org, timesheet_id, caller.Sub, reason, cancellationToken);
    }
}
